src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -91,6 +91,76 @@
 EntitiesByType = Dict[EntityType, Set[EntityID]]
 
 
+@attr.s(frozen=True, auto_attribs=True, kw_only=True)
+class Links:
+    project: EntityReference
+    processes: Entities
+    protocols: Entities
+    inputs: Entities
+    outputs: Entities
+    supplementary_files: Entities
+
+    @classmethod
+    def for_project(cls, project: EntityReference) -> 'Links':
+        return cls(project=project,
+                   **{f.name: set() for f in attr.fields(cls) if f.name != 'project'})
+
+    @classmethod
+    def from_json(cls, links_json: JSONs, project: EntityReference) -> 'Links':
```

Comment:
The higher level entity should be passed first. We usually order arguments from general (left) to specific (right).